File: 05 Bagels/java/Bagels.java
/******************************************************************************
*
* Bagels
*
* From: BASIC Computer Games (1978)
*       Edited by David H. Ahl
*
* "In this game, the computer picks a 3-digit secret number using
*  the digits 0 to 9 and you attempt to guess what it is.  You are
*  allowed up to twenty guesses.  No digit is repeated.  After
*  each guess the computer will give you clues about your guess
*  as follows:
*
*  PICO     One digit is correct, but in the wrong place
*  FERMI    One digit is in the correct place
*  BAGELS   No digit is correct
*
* "You will learn to draw inferences from the clues and, with
*  practice, you'll learn to improve your score.  There are several
*  good strategies for playing Bagels.  After you have found a good
*  strategy, see if you can improve it.  Or try a different strategy
*  altogether and see if it is any better.  While the program allows
*  up to twenty guesses, if you use a good strategy it should not
*  take more than eight guesses to get any number.
*
* "The original authors of this program are D. Resek and P. Rowe of
*  the Lawrence Hall of Science, Berkeley, California."
*
* Java port by Jeff Jetton, 2020, based on an earlier Python port
*
******************************************************************************/

import java.util.Scanner;

public class Bagels {

  public static void main(String[] args) {
    
    int gamesWon = 0;

    // Intro text
    System.out.println("\n\n                Bagels");
    System.out.println("Creative Computing  Morristown, New Jersey");
    System.out.println("\n\n");
    System.out.print("Would you like the rules (Yes or No)? ");
    
    // Need instructions?
    Scanner scan = new Scanner(System.in);
    String s = scan.nextLine();
    if (s.length() == 0 || s.toUpperCase().charAt(0) != 'N') {
      System.out.println();
      BagelInstructions.printInstructions();
    }
    
    // Loop for playing multiple games
    boolean stillPlaying = true;
    while(stillPlaying) {
      
      // Set up a new game
      BagelGame game = new BagelGame();
      System.out.println("\nO.K.  I have a number in mind.");
    
      // Loop guess and responsses until game is over
      while (!game.isOver()) {
        String guess = getValidGuess(game);
        String response = game.makeGuess(guess);
        // Don't print a response if the game is won
        if (!game.isWon()) {
          System.out.println(response);
        }
      }
    
      // Game is over. But did we win or lose?
      if (game.isWon()) {
        System.out.println("You got it!!!\n");
        gamesWon++;
      } else {
        System.out.println("Oh well");
        System.out.print("That's " + BagelGame.MAX_GUESSES + " guesses.  ");
        System.out.println("My number was " + game.getSecretAsString());
      }
      
      stillPlaying = getReplayResponse();
    }
    
    // Print goodbye message
    if (gamesWon > 0) {
      System.out.println("\nA " + gamesWon + " point Bagels buff!!");
    }
    System.out.println("Hope you had fun.  Bye.\n");
  }
  
  private static String getValidGuess(BagelGame game) {
    // Keep asking for a guess until valid
    Scanner scan = new Scanner(System.in);
    boolean valid = false;
    String guess = "";
    String error;
    
    while (!valid) {
      System.out.print("Guess # " + game.getGuessNum() + "     ? ");
      guess = scan.nextLine().trim();
      error = game.validateGuess(guess);
      if (error == "") {
        valid = true;
      } else {
        System.out.println(error);
      }
    }
    return guess;
  }
  
  private static boolean getReplayResponse() {
    // keep asking for response until valid
    Scanner scan = new Scanner(System.in);
    // Keep looping until a non-zero-length string is entered
    while (true) {
      System.out.print("Play again (Yes or No)? ");
      String response = scan.nextLine().trim();
      if (response.length() > 0) {
        return response.toUpperCase().charAt(0) == 'Y';
      }
    }
  }
  
}

File: 05 Bagels/ruby/bagels.rb
# Bagels
# Number guessing game.
# Original source unknown but suspected to be
# Lawrence Hall of Science, U.C. Berkely

def print_instructions
  puts
  puts 'I am thinking of a three-digit number. Try to guess'
  puts 'my number and i will give you clues as follows:'
  puts '   PICO   - one digit correct but in the wrong position'
  puts '   FERMI  - one digit correct and in the right position'
  puts '   BAGELS - no digits correct'
end

def generate_number_array
  target = 0
  target = rand(100..999) while target.to_s.split('').uniq.size < 3
  target.to_s.split('')
end

def puts_clue_for(guess_array, target_array)
  clues = [].tap do |clue|
    guess_array.each_with_index do |n, i|
      if target_array[i] == n
        clue << 'FERMI'
      elsif target_array.include?(n)
        clue << 'PICO'
      end
    end
  end

  # sort clues so that FERMIs come before PICOs, but
  # you don't know which response responds to which number
  puts clues.length > 0 ? clues.sort.join(' ') : 'BAGELS'
end

player_points = 0
desire_to_play = true

puts 'Bagels'.center(72)
puts 'Creative Computing  Morristown, New Jersey'.center(72)

5.times { puts }

puts 'Would you like to the rules? [Y]es or [N]o.'
instructions_request = gets.chomp.downcase

print_instructions if %w[yes y].include?(instructions_request)

while desire_to_play
  target_number_array = generate_number_array

  2.times { puts }
  puts 'OK. I have a number in mind.'

  guess_count = 0
  guess_array = []

  while (guess_array != target_number_array) && guess_count < 20
    guess_count += 1

    puts "Guess ##{guess_count}:"

    guess = gets.chomp

    if guess =~ /[^1-9]/
      puts 'What?'
      next
    end

    if guess.length != 3
      puts 'Try guessing a three digit number'
      next
    end

    guess_array = guess.split('')

    if guess_array.uniq.size < 3
      puts 'Oh, I forgot to tell you: the number I have in mind has no two digits the same.'
      next
    end

    if guess_array != target_number_array
      puts_clue_for(guess_array, target_number_array)
      puts
    end
  end

  if guess_array == target_number_array
    player_points += 1

    puts 'You got it!!!'
    puts
  else
    puts 'Oh well.'
    puts "That's twenty guesses. My number was #{target_number_array.join('')}."
  end

  puts
  puts 'Would you like to play again? [Y]es or [N]o'
  play_again_request = gets.chomp
  desire_to_play = %w[yes y].include?(play_again_request)
end

puts "A #{player_points} point bagels buff!!"
puts 'Hope you had fun. Bye.'
